WebApps UI for grouped search fields

MartinRiese opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary

* UI Story: https://dimagi-dev.atlassian.net/browse/USH-3903
* Spec: https://docs.google.com/document/d/1syEA-HC5g1FOjZ6m8DlfBspzKn9hKvq0zbMNbIWzQVE
* Design doc: https://docs.google.com/document/d/1ZkN0_ta8jEjQ5IYHLIduBN6ZFcVEjfSIQrN3AQgBVgw
* [HQ changes PR](https://github.com/dimagi/commcare-hq/pull/33828)
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations
- This is part of a series of PR

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

MartinRiese committed:
Add function to group search terms

MartinRiese committed:
show groups

MartinRiese committed:
extract getting children

MartinRiese committed:
collapsible group

MartinRiese committed:
fix reload

MartinRiese committed:
Fix table and make collapsible

MartinRiese committed:
Lint

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/controller.js:
```diff
@@ -104,6 +104,34 @@ hqDefine("cloudcare/js/formplayer/menus/controller", function () {
+        };
+
+        displays.forEach(display => {
+            const groupKey = typeof display.get === 'function' ?  display.get('groupKey') : display['groupKey'];
```
Not sure if I can do something better here. In one place the response object is passed in and in another a normal object.

nospame replied:
Does `_.get(object, path)` work here? I think underscore object methods often also work with things that are "object-like"

MartinRiese replied:
getting this:
```
jQuery.Deferred exception: required is not defined anonymous@http://localhost:8000/static/underscore/underscore-min.js line 6 > Function:12:6
f@http://localhost:8000/static/underscore/underscore-min.js:6:15311
render@http://localhost:8000/static/marionette.templatecache/lib/marionette.templatecache.min.js:12:620
_renderTemplate@http://localhost:8000/static/backbone.marionette/lib/backbone.marionette.js:595:23
render@http://localhost:8000/static/backbone.marionette/lib/backbone.marionette.js:1884:12
```

MartinRiese commented:
@stephherbers  @nospame  I mocked the formplayer response in `riese/mock_search_groups`

stephherbers commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/collections.js:
```diff
@@ -73,6 +73,10 @@ hqDefine("cloudcare/js/formplayer/menus/collections", function () {
             'session_id',
         ],
 
+        queryProperties: [
```
this would not exist in the check on line 158 in `controller.js` because that would be checking for an entities  response-- so this would need to be available for query and entities responses

MartinRiese replied:
I made some changes so the lines are shifted but I think you mean these line:
```
if (Object.keys(queryResponse.groupHeaders).length > 0) {
    queryCollection = new Collection(menusUtils.groupDisplays(queryResponse.displays, queryResponse.groupHeaders));
}
```
`queryResponse` is part of the entities response and does get copied. So this works in both cases.

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -661,15 +738,16 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
                 if (response.queryResponse) {
                     _.each(response.queryResponse.displays, function (responseModel, i) {
-                        self.collection.models[i].set({
+                        self._getChildModels().set({
```
Is this `.set` is working on a group of models now instead of an individual one? (and does that also mean `i` from the line above is unused?)

MartinRiese replied:
That was supposed to be `self._getChildModels()[i].set({`

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/spec/controller_spec.js:
```diff
@@ -0,0 +1,86 @@
+                        ],
+                    },
+                ]
+                console.log(grouped);
```
should be removed?

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -597,9 +598,9 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
 
-                        self.children.findByIndex(i)._setItemset(choices, response.models[i].get('itemsetChoicesKey'));
+                        self._getChildren().findByIndex(i)._setItemset(choices, response.models[i].get('itemsetChoicesKey'));
 
-                        self.children.findByIndex(i)._render();      // re-render with new choice values
+                        self._getChildren().findByIndex(i)._render();      // re-render with new choice values
```
nit: I'm sure this will get optimized away but is it possible to use `children` here rather than calling `_getChildren` multiple times?

MartinRiese replied:
No, because `.children` does not include nested children. But I pulled `self._getChildren().findByIndex(i)` out into a variable.

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -610,7 +611,7 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
         clearAction: function () {
             var self = this;
-            this.children.each(function (childView) {
+            this._getChildren().forEach(function (childView) {
```
nit: see it was already like this, but can this use `self` instead of `this` here for consistency?

nospame commented on corehq/apps/cloudcare/templates/formplayer/query_view.html:
```diff
@@ -40,7 +40,20 @@ <h2 tabindex="0"><%- title %></h2>
+<script type="text/template" id="query-view-group-template">
+  <td>
+    <div data-toggle="collapse" data-target="#group-<%- groupKey %>" aria-expanded="true" aria-controls="group-<%- groupKey %>">
+      <label><%- groupName %></label>
```
I don't know if `<label>` is relevant inside a div, as far as I know it's only for labeling an input element. And it already serves as the "label" for a clickable div by being contained inside it.

MartinRiese replied:
I mainly used it for styling since the fields themselves use it too.

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -449,10 +449,10 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
         template: _.template($("#query-view-group-template").html() || ""),
         childView: QueryView,
-        childViewContainer: "#query-properties",
+        childViewContainer: "#group-content",
```
nit: It might be nice to have something more explicitly named since it's already the page already has some elements that use "content". Something like `query-group-content` maybe?

MartinRiese replied:
sure thing

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/controller.js:
```diff
@@ -128,7 +156,10 @@ hqDefine("cloudcare/js/formplayer/menus/controller", function () {
         if (sidebarEnabled && menuResponse.type === "entities" && queryResponse)  {
-            var queryCollection = new Collection(queryResponse.displays);
+            let queryCollection = new Collection(queryResponse.displays);
+            if (queryResponse.hasOwnProperty("groupHeaders")) {
```
I might have told you the wrong thing before. But `groupHeaders` will always be in the response. If there's no groups, the value will be `{}` so you'll need to also check if it's empty here and below

MartinRiese replied:
replacing it with `Object.keys(queryResponse.groupHeaders).length > 0`

MartinRiese committed:
Merge branch 'master' into riese/search_group

MartinRiese committed:
Address review comments

MartinRiese committed:
Groups for non sscs

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/controller.js:
```diff
@@ -157,8 +129,8 @@ hqDefine("cloudcare/js/formplayer/menus/controller", function () {
             let queryCollection = new Collection(queryResponse.displays);
-            if (queryResponse.hasOwnProperty("groupHeaders")) {
-                queryCollection = new Collection(groupDisplays(queryResponse.displays, queryResponse.groupHeaders));
+            if (Object.keys(queryResponse.groupHeaders).length > 0) {
```
I'm not entirely sure, but do we still need to first check `queryResponse` contains `groupHeaders`? I think there would be a transient moment during deploy where FP changes aren't deployed yet so queryResponse doesn't contain `groupHeaders` and this would throw an error.

I think ideally, formplayer would be deployed first and then HQ the next week but it's most likely happening at the same time for this feature.

MartinRiese replied:
Good. Point I will put that back in

MartinRiese committed:
Fix styling

MartinRiese committed:
more styling

MartinRiese committed:
Merge branch 'master' into riese/search_group

MartinRiese committed:
Fix if check

MartinRiese committed:
Fix expandy arrow behavior

MartinRiese committed:
Fix expandy arrow behavior

MartinRiese committed:
Add description

MartinRiese committed:
Mark groups as required if one field is

MartinRiese committed:
Unnamed groups do not collapse

MartinRiese committed:
Set white bg when not grouped and no SSCS

Jtang-1 committed:
bug: converts display array into backbone models

Jtang-1 committed:
bug: checks that property exists before calling

there would be a transient moment during deploy where FP changes aren't deployed yet so
response does not contain groupHeaders property

Jtang-1 committed:
bug: _getChildren returns array so accessing item should be done as such

Jtang-1 committed:
bug: gets the base level model for each query question

child of querylistview can be either queryView which represents a query field
or GroupedQueryView which represents a group of query fields

Jtang-1 committed:
feat: allow markup for group name header text

MartinRiese committed:
Merge branch 'master' into riese/search_group

MartinRiese committed:
Fix linting errors

Jtang-1 committed:
bug: missed variable rename

MartinRiese committed:
Rename spec and update import

MartinRiese committed:
Add group headers to mock response

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/controller.js:
```diff
@@ -128,7 +128,11 @@ hqDefine("cloudcare/js/formplayer/menus/controller", function () {
-            var queryCollection = new Collection(queryResponse.displays);
+            let queryCollection = new Collection(queryResponse.displays);
+            // hasOwnProperty check only required temporarily for transient moment during deploy when response will not contain groupHeaders
+            if (queryResponse.hasOwnProperty("groupHeaders") && Object.keys(queryResponse.groupHeaders).length > 0) {
```
I was not able to update the collection inside the view. but I will give it one more try

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -460,10 +460,74 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
+        },
+
+        onAttach: function () {
+            const header = $(this.el).find('.search-query-group-header');
```
ok, then would you consider putting header into `ui`: and have the behavior be an `event` similar to these pattern we use here? https://github.com/dimagi/commcare-hq/blob/15e891f5800b058b39971b32b123c7db8c318776/corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js#L558

MartinRiese replied:
good point https://github.com/dimagi/commcare-hq/pull/33805/commits/03726eea4b2a4ce73f887ea73cb364bfacca5dda

MartinRiese committed:
Fix failing tests

MartinRiese committed:
Lint

MartinRiese committed:
Merge branch 'master' into riese/search_group

MartinRiese committed:
Use ui and events instead of jquery

MartinRiese committed:
Move logic for grouped search into QueryListView

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/controller.js:
```diff
@@ -128,35 +128,38 @@ hqDefine("cloudcare/js/formplayer/menus/controller", function () {
-            if (queryResponse.hasOwnProperty("groupHeaders") && Object.keys(queryResponse.groupHeaders).length > 0) {
-                queryCollection = new Collection(menusUtils.groupDisplays(new Collection(queryResponse.displays), queryResponse.groupHeaders));
+            let myGroupHeaders = {};
+            if (queryResponse.hasOwnProperty("groupHeaders")) {
```
I ended up deploying Formplayer to prod already so we will no longer need these checks.

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -827,7 +871,10 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
+        queryListView: function (data) {
+            return new QueryListView(data);
+        },
+        groupDisplays: groupDisplays,
```
I think the testing you added for this is great. I'm seeing this is added for just the test? It might just be a personal preference so this is a non-blocking comment, maybe we can test this on a view level like what was done in `query_spec.js`

Jtang-1 reviewed (commented):
couple additional comments

MartinRiese committed:
Fix linting and test

MartinRiese committed:
Remove not needed code

MartinRiese committed:
import backbone

MartinRiese committed:
Merge branch 'master' into riese/search_group

MartinRiese committed:
fix spec

Jtang-1 reviewed (approved):
Looks good, but there are some non-blocking comments from looking at it again

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -460,10 +496,80 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
+        buildChildView(child, ChildViewClass, childViewOptions) {
+            let options = {};
+
+            if (child.has("groupName")) {
```
No. that is to identify if the child is a group as set in `groupDisplays`

Jtang-1 replied:
But if a group's `groupName` is an empty string, then wouldn't this incorrectly evaluate as false? 

If we check `groupKey`'s value instead, then when it's null, it wouldn't be a group. Otherwise, if `groupKey` contains a value, it is a group.

MartinRiese replied:
This is to distinguish between showing groups or not. If `groupHeaders` is empty the `displays` will be used as they are (and they do not have a field `groupName`. If there are group headers the `collection` will be "group" objects that have a `groupName` field

Jtang-1 replied:
oh i see, you're right. I mistook that it was checking the value of `groupName` but it's just checking if the property exists.

MartinRiese committed:
fix some tests

MartinRiese committed:
fix test input to be collection

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/spec/fixtures/split_screen_case_list.js:
```diff
@@ -34,6 +34,9 @@ hqDefine("cloudcare/js/formplayer/spec/fixtures/split_screen_case_list", functio
                 "details": null,
             },
         ],
+        "groupHeaders": {
```
this should be inside `queryResponse`

MartinRiese replied:
I just put it in both because it can be in either

Jtang-1 replied:
What do you mean by that? It isn't directly in the entity response. It would only be nested the `queryResponse` and the `queryResponse` would be added to the entity response.

Jtang-1 reviewed (changes requested):
It doesn't work as is, it errors when i test it locally since there's no `groupHeaders` in entity type `menuResponse`. 

Making these changes breaks the tests again.. I think i figured out what the issue is. See here for fixing the broken tests https://github.com/dimagi/commcare-hq/compare/riese/search_group...jt/search_group_test_fix?expand=1

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/controller.js:
```diff
@@ -135,7 +135,7 @@ hqDefine("cloudcare/js/formplayer/menus/controller", function () {
                     sidebarEnabled: true,
                     disableDynamicSearch: !sessionStorage.submitPerformed,
-                    groupHeaders: queryResponse.groupHeaders,
+                    groupHeaders: menuResponse.groupHeaders,
```
you are right.

MartinRiese committed:
moving this back

MartinRiese committed:
check trimmed group names size since empty group names come back as " "

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -460,10 +496,80 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
+            let groupName = this.options.groupName === undefined ?
+                "" : markdown.render(this.options.groupName.trim());
+            return {
+                groupName: DOMPurify.sanitize(groupName),
```
[bc6620e](https://github.com/dimagi/commcare-hq/pull/33805/commits/bc6620ee9c11d573f141322cf2aff7f4da12a859)

MartinRiese committed:
check trimmed group names size since empty group names come back as " " part 2

forgot the length

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -460,10 +496,80 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
+                groupName: DOMPurify.sanitize(groupName),
+                groupKey: this.options.groupKey,
+                required: this.options.required,
+                named: this.options.groupName.length > 0,
```
[2646960](https://github.com/dimagi/commcare-hq/pull/33805/commits/264696077ddf5ef4a9aeea0c6ef7b9477d22b980)

esoergel reviewed (approved):
I have not reviewed this PR, but I'm reasserting Jonathan's approval, since github seems not to consider it satisfactory.  I believe that we have met our team's review standards for merge.

![image](https://github.com/dimagi/commcare-hq/assets/2367539/5f8c5cd4-9d96-4cec-9575-f2931d7f9fb1)
